### ISSUE TITLE
Cannot build with CUDA/cuDNN on OSX

### ISSUE THREAD

**Issue by kylemcdonald** [davisking/dlib#289]
On OSX 10.11 with Xcode 7.3, CUDA 8.0 and cuDNN 5.0.5 first I run:

```
$ cd ~/Desktop ; git clone https://github.com/davisking/dlib.git
$ cd dlib ; mkdir build ; cd build
$ cmake ..
```

And the output is:

```
-- The C compiler identification is AppleClang 8.0.0.8000038
-- The CXX compiler identification is AppleClang 8.0.0.8000038
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- C++11 activated.
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE  
-- Looking for png_create_read_struct
-- Looking for png_create_read_struct - found
-- Looking for jpeg_read_header
-- Looking for jpeg_read_header - found
-- Searching for BLAS and LAPACK
-- Found PkgConfig: /usr/local/bin/pkg-config (found version "0.29.1") 
-- Checking for module 'cblas'
--   No package 'cblas' found
-- Checking for module 'lapack'
--   No package 'lapack' found
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of void*
-- Check size of void* - done
-- Found Intel MKL BLAS/LAPACK library
-- Found CUDA: /Developer/NVIDIA/CUDA-8.0 (found suitable version "8.0", minimum required is "7.5") 
-- Looking for cuDNN install...
-- Building a CUDA test project to see if your compiler is compatible with CUDA...
-- *** CUDA was found but your compiler failed to compile a simple CUDA program so dlib isn't going to use CUDA. ***
-- *** cuDNN V5.0 OR GREATER NOT FOUND.  DLIB WILL NOT USE CUDA. ***
-- *** If you have cuDNN then set CMAKE_PREFIX_PATH to include cuDNN's folder.
-- Configuring done
CMake Warning (dev):
  Policy CMP0042 is not set: MACOSX_RPATH is enabled by default.  Run "cmake
  --help-policy CMP0042" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  MACOSX_RPATH is not specified for the following targets:

   dlib_shared

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /Users/kyle/Desktop/dlib/build
```

I try starting over with additional settings:

```
$ rm -rf *
$ cmake .. -DCMAKE_MACOSX_RPATH=1 -DCMAKE_PREFIX_PATH=/Developer/NVIDIA/CUDA-8.0
```

But the result is the same (except for the note about MACOSX_RPATH). To be clear:

```
$ ls /usr/local/cuda/lib/libcud*
/usr/local/cuda/lib/libcuda.dylib       /usr/local/cuda/lib/libcudart_static.a
/usr/local/cuda/lib/libcudadevrt.a      /usr/local/cuda/lib/libcudnn.5.dylib
/usr/local/cuda/lib/libcudart.8.0.dylib /usr/local/cuda/lib/libcudnn.dylib
/usr/local/cuda/lib/libcudart.dylib     /usr/local/cuda/lib/libcudnn_static.a
$ ls /usr/local/cuda/include/cud*
/usr/local/cuda/include/cuda.h
...
/usr/local/cuda/include/cudnn.h
$ ls /usr/local/cuda/bin/nvcc
/usr/local/cuda/bin/nvcc
```

And results are the same for `/Developer/NVIDIA/CUDA-8.0`, `/usr/local/cuda` is mostly making aliases to there.

```
$ xcode-select -p
/Applications/Xcode7.3.app/Contents/Developer
```

I'm using CUDA and cuDNN with other tools, like keras. So I'm not sure why it's not working here. Any ideas?

**Comment by e-fominov** [comment]
> -- *** CUDA was found but your compiler failed to compile a simple CUDA program so dlib isn't going to use CUDA. ***

This means CUDA sample failed to build. Sample is located in dlib/cmake_utils/test_for_cuda and it has CMakeLists.txt inside

Try to build this sample manually first;
```
mkdir build
cd build 
cmake ../dlib/cmake_utils/test_for_cuda
cmake --build .
```
 If it fails to build - you will get error messages and then you can understand what is wrong. The next sample is CUDNN sample and it is in  dlib/cmake_utils/test_for_cudnn - same situation

**Comment by kylemcdonald** [comment]
thanks a ton! the first thing i found is:

```
$ cmake ..
-- The C compiler identification is AppleClang 7.3.0.7030031
-- The CXX compiler identification is AppleClang 7.3.0.7030031
-- Check for working C compiler: /Applications/Xcode7.3.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc
-- Check for working C compiler: /Applications/Xcode7.3.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Applications/Xcode7.3.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++
-- Check for working CXX compiler: /Applications/Xcode7.3.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE  
-- Found CUDA: /Developer/NVIDIA/CUDA-8.0 (found suitable version "8.0", minimum required is "7.5") 
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/kyle/Desktop/dlib/dlib/cmake_utils/test_for_cuda/build
$ cmake --build .
[ 50%] Building NVCC (Device) object CMakeFiles/cuda_test.dir/cuda_test_generated_cuda_test.cu.o
In file included from <built-in>:356:
In file included from <command line>:15:
In file included from /Developer/NVIDIA/CUDA-8.0/include/cuda_runtime.h:116:
/Developer/NVIDIA/CUDA-8.0/include/common_functions.h:65:10: fatal error: 'string.h' file not found
#include <string.h>
         ^
1 error generated.
CMake Error at cuda_test_generated_cuda_test.cu.o.cmake:207 (message):
  Error generating
  /Users/kyle/Desktop/dlib/dlib/cmake_utils/test_for_cuda/build/CMakeFiles/cuda_test.dir//./cuda_test_generated_cuda_test.cu.o


make[2]: *** [CMakeFiles/cuda_test.dir/cuda_test_generated_cuda_test.cu.o] Error 1
make[1]: *** [CMakeFiles/cuda_test.dir/all] Error 2
make: *** [all] Error 2
```

googling this lead me to this post https://github.com/torch/cutorch/issues/241#issuecomment-146628841

when xcode updates, the command line tools are uninstalled, and you need to run `xcode-select --install` again. i'm not sure how other things were compiling without the command line tools.

after running `xcode-select --install` the `test_for_cuda` compiles without problems.

then i checked `test_for_cudnn` and got this:

```
$ cmake --build .
Scanning dependencies of target cudnn_test
[ 50%] Building CXX object CMakeFiles/cudnn_test.dir/Users/kyle/Desktop/dlib/dlib/dnn/cudnn_dlibapi.cpp.o
/Users/kyle/Desktop/dlib/dlib/dnn/cudnn_dlibapi.cpp:115:13: error: thread-local storage is not supported for the current target
            thread_local cudnn_context c;
            ^
/Users/kyle/Desktop/dlib/dlib/dnn/cudnn_dlibapi.cpp:154:13: error: thread-local storage is not supported for the current target
            thread_local cudnn_activation_descriptor des(CUDNN_ACTIVATION_RELU, CUDNN_PROPAGATE_NAN,0);
            ^
/Users/kyle/Desktop/dlib/dlib/dnn/cudnn_dlibapi.cpp:160:13: error: thread-local storage is not supported for the current target
            thread_local cudnn_activation_descriptor des(CUDNN_ACTIVATION_SIGMOID, CUDNN_PROPAGATE_NAN,0);
            ^
/Users/kyle/Desktop/dlib/dlib/dnn/cudnn_dlibapi.cpp:166:13: error: thread-local storage is not supported for the current target
            thread_local cudnn_activation_descriptor des(CUDNN_ACTIVATION_TANH, CUDNN_PROPAGATE_NAN,0);
            ^
4 errors generated.
make[2]: *** [CMakeFiles/cudnn_test.dir/Users/kyle/Desktop/dlib/dlib/dnn/cudnn_dlibapi.cpp.o] Error 1
make[1]: *** [CMakeFiles/cudnn_test.dir/all] Error 2
make: *** [all] Error 2
```

the solution to this issue is described in this thread https://github.com/davisking/dlib/issues/223 and fixed by this branch https://github.com/grefab/dlib/commits/feature/osx_cuda_build

after switching to that branch everything compiles, but then when i run the face detection tool on all the faces in the "faces" directory, after a few faces have been processed i see: `Error while calling cudaMalloc(&forward_workspace, forward_workspace_size_in_bytes) in file /Users/kyle/Desktop/dlib/dlib/dnn/cudnn_dlibapi.cpp:889. code: 2, reason: out of memory` this is reported after a different number of faces each time i run it, so it's not one specific image size that's causing a problem...

but regardless, this specific issue is now closed :) thanks!

**Comment by kylemcdonald** [comment]
i just learned from https://github.com/arrayfire/arrayfire/issues/1384#issuecomment-265797045 and the [CUDA release notes] the latest CUDA 8.0.55 supports clang 3.7 and 3.8 (XCode 7 and 8). just tested and it works for me on OSX 10.11 XCode 8.2.1
